### PR TITLE
Use retry switching peer for world state download tasks

### PR DESCRIPTION
**Pull request by fab-10**
## PR description

Built on top of #5509, so please check it first. [Link to only see diff again #5509](https://github.com/fab-10/besu/compare/demote-peer-with-empty-responses-when-retrying...fab-10:besu:use-retry-switching-peer)

With this PR all the world state download tasks used by snap sync are updated to use the retry switching peer strategy, to reduce the chance that we keep sending requests to the same bad peer.

relates to https://github.com/hyperledger/besu/issues/5415 and #5271

### CODE REVIEW

File: ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/manager/snap/RetryingGetStorageRangeFromPeerTask.java [hunk 1]
```diff
@@ -44,8 +43,13 @@ private RetryingGetStorageRangeFromPeerTask(
...
       final Bytes32 endKeyHash,
       final BlockHeader blockHeader,
-      final MetricsSystem metricsSystem) {
-    super(ethContext, 4, data -> data.proofs().isEmpty() && data.slots().isEmpty(), metricsSystem);
+      final MetricsSystem metricsSystem,
+      final int maxRetries) {
+    super(
+        ethContext,
+        metricsSystem,
+        data -> data.proofs().isEmpty() && data.slots().isEmpty(),
```
Comment (matkt) [inline ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/manager/snap/RetryingGetStorageRangeFromPeerTask.java:51]: A bad peer will be a peer with this kind of result
I’m not sure you are checking if the response is like that in order to switch to another peer

Comment (fab-10) [reply ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/manager/snap/RetryingGetStorageRangeFromPeerTask.java:51]: check this other PR https://github.com/hyperledger/besu/pull/5509, that checks and records the empty response in the `AbstractRetryingPeerTask`

File: ethereum/eth/src/test/java/org/hyperledger/besu/ethereum/eth/manager/ethtaskutils/RetryingMessageTaskTest.java [hunk 2]
```diff
@@ -155,7 +166,8 @@ public void completesWhenPeersAreTemporarilyUnavailable()
     final T requestedData = generateDataToBeRequested();
 
     // Execute task and wait for response
-    final EthTask<T> task = createTask(requestedData);
+    // +2 is required for switching peers tasks where the max retries = number of peers
```
Comment (siladu) [inline ethereum/eth/src/test/java/org/hyperledger/besu/ethereum/eth/manager/ethtaskutils/RetryingMessageTaskTest.java:169]: useful comment 👍

File: ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/manager/EthPeers.java [hunk 3]
```diff
@@ -320,7 +320,9 @@ public Stream<EthPeer> streamAvailablePeers() {
   }
 
   public Stream<EthPeer> streamBestPeers() {
-    return streamAvailablePeers().sorted(getBestChainComparator().reversed());
+    return streamAvailablePeers()
+        .filter(EthPeer::isFullyValidated)
```
Comment (siladu) [inline ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/manager/EthPeers.java:324]: Why is this required?

**Review by siladu — COMMENTED** [review]
Diff looks pretty straightforward. Couple of small comments. Don't want to approve while the parent PR is in flight

**Comment by matkt** [comment]
Just a question. A peer that is useless for the worldstate part may not be for the blockchain part. Is there not a risk of losing valid nodes for a certain type of data by marking useless when it could still help us.

**Comment by macfarla** [comment]
Blocked by #6609 - without that we would disconnect peers, because we are trying peers that do not serve snap data.

**Comment by jframe** [comment]
@fab-10 Closing this, reopen if think we still need it
